### ISSUE TITLE
Deeplinks not working from push notification action buttons on iOS when app is closed

### ISSUE THREAD

**Issue by jwebcat** [Appboy/appboy-react-sdk#103]
I integrated push notification, deeplinks and default categories for the action buttons on iOS following the demo app in this repo and iOS push notification integration docs. 

##### When opening the app from the push notification action buttons with deeplink configured:

- If the app is closed, the `ReactAppboy.getInitialURL` handler is not being called. After digging a bit into the integration,  the `populateInitialUrlFromLaunchOptions` method in `AppboyReactUtils.m` file is being called but the `launchOptions` are nil.

- If the app is in the background, the deeplinks are working from the action buttons as expected.

##### When opening the app from the push notification body with deeplink configured:

- it's working as expected with the app in the background and fully closed.

I was able to replicate this issue in the AppboyProject demo app as well. Any insight / help on this issue would be much appreciated.

**Comment by Bucimis** [comment]
@jwebcat thanks for letting us know. Quick initial question, are you using user notification center delegates or only application delegates?

**Comment by jwebcat** [comment]
@Bucimis Thanks for the quick reply. I'm using user notification center delegates and remote notification background mode is enabled.

**Comment by Bucimis** [comment]
@jwebcat thanks, that'll help us investigate. We're going to pick this up next week, but the issue is that action button support isn't implemented. Next steps for that are to:
1) Find the notification payload for action button push
2) Parse the onClick uri for the correct action button 
3) Set the uri on our `AppboyReactUtils` instance before `getInitialUrl` is called (`populateInitialUrlFromLaunchOptions` basically does step 2 & 3 for body clicks today)

**Comment by jwebcat** [comment]
@Bucimis Of course, np. Ok cool, that sounds great. Thanks for responding so quickly.

**Comment by Bucimis** [comment]
@jwebcat Could you test adding the following code to your local `AppboyReactUtils.h/m` and `UNUserNotificationCenterDelegate`?

```
// add to AppboyReactUtils.h
- (BOOL)populateInitialUrlForCategories:(NSDictionary *)userInfo;
```

```
// add to AppboyReactUtils.m
- (BOOL)populateInitialUrlForCategories:(NSDictionary *)userInfo {
  // When action buttons are opened from a cold start, didFinishLaunchingWithOptions's launchOptions are nil.
  if (sharedInstance.initialUrlString) {
    NSLog(@"initialUrlString already populated in didFinishLaunchingWithOptions. Doing nothing.");
    return false;
  }
  NSDictionary *categories = [userInfo valueForKeyPath:@"ab.ab_cat"];
  if (categories && [categories count] > 0) {
    NSDictionary *category = [[categories allValues] objectAtIndex:0];
    if (category[@"a_uri"]) {
      sharedInstance.initialUrlString = category[@"a_uri"];
      RCTLogInfo(@"[AppboyReactUtils sharedInstance].initialUrlString set to %@.", sharedInstance.initialUrlString);
      return true;
    }
  }
  sharedInstance.initialUrlString = nil;
  return false;
}
```

```
// add to the beginning of userNotificationCenter:didReceiveNotificationResponse:withCompletionHandler:
[[AppboyReactUtils sharedInstance] populateInitialUrlForCategories:response.notification.request.content.userInfo];
```

**Comment by jwebcat** [comment]
@Bucimis I added the above and tested it out, it's working like a charm 🎉 

Thanks for fixing this so quickly. I will create a patch for now, so we can test internally a bit more. Any idea when this fix might be added to the library?

**Comment by jwebcat** [comment]
Adding the patch here, in case it's useful to anyone.

`react-native-appboy-sdk+1.26.0.patch`

```
diff --git a/node_modules/react-native-appboy-sdk/iOS/AppboyReactBridge/AppboyReactBridge/AppboyReactUtils.h b/node_modules/react-native-appboy-sdk/iOS/AppboyReactBridge/AppboyReactBridge/AppboyReactUtils.h
index 05b99db..f7755e6 100644
--- a/node_modules/react-native-appboy-sdk/iOS/AppboyReactBridge/AppboyReactBridge/AppboyReactUtils.h
+++ b/node_modules/react-native-appboy-sdk/iOS/AppboyReactBridge/AppboyReactBridge/AppboyReactUtils.h
@@ -5,6 +5,7 @@
 
 + (AppboyReactUtils *)sharedInstance;
 - (BOOL)populateInitialUrlFromLaunchOptions:(NSDictionary *)launchOptions;
+- (BOOL)populateInitialUrlForCategories:(NSDictionary *)userInfo;
 
 @property NSString *initialUrlString;
 
diff --git a/node_modules/react-native-appboy-sdk/iOS/AppboyReactBridge/AppboyReactBridge/AppboyReactUtils.m b/node_modules/react-native-appboy-sdk/iOS/AppboyReactBridge/AppboyReactBridge/AppboyReactUtils.m
index f19730f..59d340b 100644
--- a/node_modules/react-native-appboy-sdk/iOS/AppboyReactBridge/AppboyReactBridge/AppboyReactUtils.m
+++ b/node_modules/react-native-appboy-sdk/iOS/AppboyReactBridge/AppboyReactBridge/AppboyReactUtils.m
@@ -30,4 +30,23 @@ - (BOOL)populateInitialUrlFromLaunchOptions:(NSDictionary *)launchOptions {
   return false;
 }
 
+- (BOOL)populateInitialUrlForCategories:(NSDictionary *)userInfo {
+  // When action buttons are opened from a cold start, didFinishLaunchingWithOptions's launchOptions are nil.
+  if (sharedInstance.initialUrlString) {
+    NSLog(@"initialUrlString already populated in didFinishLaunchingWithOptions. Doing nothing.");
+    return false;
+  }
+  NSDictionary *categories = [userInfo valueForKeyPath:@"ab.ab_cat"];
+  if (categories && [categories count] > 0) {
+    NSDictionary *category = [[categories allValues] objectAtIndex:0];
+    if (category[@"a_uri"]) {
+      sharedInstance.initialUrlString = category[@"a_uri"];
+      RCTLogInfo(@"[AppboyReactUtils sharedInstance].initialUrlString set to %@.", sharedInstance.initialUrlString);
+      return true;
+    }
+  }
+  sharedInstance.initialUrlString = nil;
+  return false;
+}
+
 @end

```

**Comment by Bucimis** [comment]
@jwebcat thanks for testing the fix and generating a patch! We'll hold our release pending the results of your additional internal testing, so let us know how that goes. Assuming no issues arise from your or our internal QA, at that point we'll release within a few days.

**Comment by jwebcat** [comment]
@Bucimis of course 🎉 

Our QA tested this and it's looking good. It was tested on the following devices:

iPhone 6s
iPhone 7+
iPhone 8
iPhone X
iPhone XR
iPhone 11 Pro Max
iPhone 12
iPhone 12 Mini
iPad 

and OS:

13.2
13.7
14.0
14.3
14.4

**Comment by Bucimis** [comment]
@jwebcat thanks for much for following up! We're working through our QA process now and are headed for release - likely in the next few days and definitely by eow unless something big crops up. Thanks again!

**Comment by jwebcat** [comment]
@Bucimis Of course, np! Thanks to you as well for getting this fixed so quickly. Sounds great, looking forward to the new release.

**Comment by wesleyorbin** [comment]
@jwebcat We just released this in 1.27.0. Thank you again for your help!

**Comment by jwebcat** [comment]
@wesleyorbin of course, glad to help! Thanks for getting this out 🎉
